### ISSUE TITLE
ha rolling-updates should only use rolled servers for API load balancer

### ISSUE THREAD

**Issue by chrislovecnm** [kubernetes/kops#1982]
When you are doing a rolling-update, with a version changes, we should have the load balancer only talk to the newer masters, once they are up.

We are not bumping the API version correctly by having 1.4 server(s) and a 1.5 server(s) listening to the same loadbalancer. Once a  master is up, only masters that are the rolled version can be listening to the load balancer.

1. roll one master
1. validate cluster
1. remove remove the other masters from LB
1. validate cluster
1. roll the next masters and add it back to the LB

Basically, you can make API call an say oh I am talking to a 1.5 cluster, then make another API call and suddenly be talking to a 1.4 cluster.

**Comment by justinsb** [comment]
Can you add a citation for this?

**Comment by chrislovecnm** [comment]
From slack with @justinsb 

> You have asserted that you can't mix master versions.  I don't think that is true, so I would like to know why you believe that to be true

1. You could be rolling a cluster from 1.4 -> 1.5 and install a petset in the middle of the roll
1. Rolling-update code uses the eviction API, and it is checking with the cluster which version of the eviction API it should be using.  It can get version a and then talk to a version b API server.
1. Any kubectl code or custom code that talks to the api server can run into nasty problems when it is doing a muliple step process:
  - check which version of a cluster I am talking to
  - make API calls with the version communicated
  - API call then goes to a different API server version

**Comment by chrislovecnm** [comment]
You could still do use case number 1, if you do it while the first master is rolling, but you could do it during all three master rolls.

**Comment by justinsb** [comment]
In general we do/should support mixed version of the kubernetes API for a relatively small time duration.  But I think the argument will be that petsets were alpha, so don't come under the guarantee.

Is the eviction API alpha/beta?  Because if not, the rules state that it must be compatible across versions - otherwise that is a bug.

**Comment by chrislovecnm** [comment]
Yep, eviction API is alpha/beta.  This will be a reoccurring theme, with API version bumps.   

> In general we do/should support mixed version of the kubernetes API for a relatively small time duration.

How?

**Comment by justinsb** [comment]
It's a kubernetes API guarantee - we don't break the API.  See https://kubernetes.io/docs/api/ which links to the API change document.

**Comment by chrislovecnm** [comment]
Well then wth does petsets go against that rule.  🙊

**Comment by justinsb** [comment]
Because petsets was alpha :-(
